lib/tracecontext/tracestate.js

```diff
@@ -0,0 +1,346 @@
+/**
+ * Class for Managing Tracestate
+ *
+ * Class that creates objects for managing trace state.
+ * This class is capable of parsing both tracestate strings
+ * and tracestate binary representations, allowing clients
+ * to get and set values in a single list-member/namespace
+ * while preserving values in the other namespaces.
+ *
+ * Capable of working with either the binary of string
+ * formatted tracestate values.
+ *
+ * Usage:
+ *   const tracestate = TraceState.fromStringFormatString(headerTracestate, 'es')
+ *   tracestate.setValue('s',1)
+ *   const newHeader = tracestate.toW3cString()
+ */
+class TraceState {
+  constructor (sourceBuffer, listMemberNamespace = 'es', defaultValues = {}, logger) {
+    if (!this._validateVendorKey(listMemberNamespace)) {
+      throw new Error('Vendor namespace failed validation.')
+    }
+
+    // buffer representation of the trace state string.
+    // The initial value of this.buffer will keep the
+    // values set in the listMemberNamespace list-member,
+    // but as soon as an initial value is set (via setValue)
+    // then the listMemberNamespace values will be removed
+    // from this.buffer and stored in the this.values.  While
+    // slightly more complicated, this allows us to maintain
+    // the order of list-member keys in an un-mutated tracestate
+    // string, per the W3C spec
+    this.buffer = sourceBuffer
+
+    this.listMemberNamespace = listMemberNamespace
+
+    // values for our namespace, set via setValue to
+    // ensure names conform
+    this.values = {}
+    for (const [key, value] of Object.entries(defaultValues)) {
+      this.setValue(key, value)
+    }
+
+    // if no logger, then fallback to agent singleton
+    logger = logger || require('../..').logger
+    this.logger = logger
+  }
+
+  setValue (key, value) {
+    const strKey = String(key)
+    const strValue = String(value)
+    if (!this._validateElasicKeyAndValue(strKey, strValue)) {
```

Comment:
Just for my understanding, Is this applicable only to elastic namespace or every other vendor key - value?

Reply:
Good question -- it's applicable only to the elastic namespace.  The intention of the `setValue` method is to allow users to set values in the elastic namespace.  The are deliberately no methods for setting values in the other namespaces (other than the initial value set when creating the tracestate object)